src/backend/parser/parse_coerce.c

```diff
@@ -1504,9 +1505,38 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 				pcategory = ncategory;
 				pispreferred = nispreferred;
 			}
+		} else if (sql_dialect == SQL_DIALECT_TSQL && ntype == ptype)
+		{
+			/*
+			 * For the columns which have the same base type, we choose the
+			 * expression with higher precedence type in T-SQL.
+			 * For example, smallmoney UNION money, the base type of
+			 * them are both fixeddecimal. But we shouldn't use smallmoney as
+			 * the result type, it could loss precision.
+			 * Here we don't need to update other variables since they are the
+			 * same.
+			 */
+			if (is_tsql_base_datatype_hook &&
+				(*is_tsql_base_datatype_hook)(exprType(nexpr)) &&
+				determine_datatype_precedence_hook &&
+				determine_datatype_precedence_hook(exprType(nexpr),
+												   exprType(pexpr)))
+				pexpr = nexpr;
```

Comment:
why arent we updating ptype? because next iteration may need to make some decision

Reply:
second, why cant we fit this logic into select_common_type_hook?

Reply:
Since in this branch, `ntype` is equal to `ptype`. So as `pcategory` and `pispreferred`. We change the `expr` for the information of domain type. This logic is combined with the following changes.

Reply:
>second, why cant we fit this logic into select_common_type_hook?

It is because a general fix for this issue should happens after the main process in `select_common_type`, in which the common type has been decided. The `select_common_type_hook` does a early return. And if we want to do it in this hook, we need to copy most codes in `select_common_type` to that hook.